File: Java/section1.4/crypt1.java
package src.section1_3;

import java.io.*;
import java.util.HashMap;

/*
ID: jameszh1
LANG: JAVA
TASK: crypt1
*/
public class crypt1 {
    public static void main (String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("crypt1.in"));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("crypt1.out")));
        HashMap<Integer, Integer> hm = new HashMap();
        int numOfDigit = Integer.parseInt(br.readLine());
        int total = 0;
        String[] line = br.readLine().split(" ");
        for (int i = 0; i < line.length; i++) {
            hm.put(Integer.parseInt(line[i]), 1);
        }

        for (int i = 100; i < 1000; i++) {
            if (isIn(i, hm)) {
                for (int j = 10; j < 100; j++) {
                    if (isIn(j, hm)) {
                        String str_j = String.valueOf(j);
                        int tenDigit = Integer.parseInt(str_j.substring(0,1));
                        int digit = Integer.parseInt(str_j.substring(1,2));
                        int middleResult1 = digit * i;
                        int middleResult2 = tenDigit * i;
                        if (isIn(middleResult1, hm) && isIn(middleResult2, hm) && isIn(i * j, hm) && i * j >= 1000 && i * j < 10000 && middleResult1 > 100 && middleResult1 < 1000 && middleResult2 > 100 && middleResult2 < 1000) {
                            total++;
                        }
                    }
                }
            }
        }

        pw.println(total);

        pw.close();
        br.close();
    }

    public static boolean isIn (int number, HashMap hm) {
        String str = String.valueOf(number);
        boolean flag = true;
        for (int i = 0; i < str.length(); i++) {
            int digit = Integer.parseInt(str.substring(i, i+1));
            if (!hm.containsKey(digit)) {
                flag = false;
                break;
            }
        }
        return flag;
    }
}

File: Python/chapter1/section4/crypt1.py
"""
ID: jameszh1
LANG: PYTHON2
TASK: crypt1
"""
def is_number_digit_in_dict(number, dictionary):
    number_str = str(number)
    for i in range(len(number_str)):
        if (not dictionary.__contains__(int(number_str[i]))):
            return False
    return True

fin = open ('crypt1.in', 'r')
fout = open ('crypt1.out', 'w')
digit_num = int(fin.readline().strip())
number_array = fin.readline().strip().split(" ")

number_dict = dict()
for i in range(len(number_array)):
    number = int(number_array[i])
    number_dict[number] = 1

total = 0
for i in range(100, 1000):
    if (is_number_digit_in_dict(i, number_dict)):
        for j in range(10, 100):
            if (is_number_digit_in_dict(j, number_dict)):
                digit_j = j % 10
                ten_digit_j = int(j / 10)
                middle_result1 = digit_j * i
                middle_result2 = ten_digit_j * i
                if (is_number_digit_in_dict(middle_result1, number_dict) and is_number_digit_in_dict(middle_result2, number_dict) and is_number_digit_in_dict(i * j, number_dict)
                       and i * j >= 1000 and i * j < 10000 and middle_result1 > 100 and middle_result1 < 1000 and middle_result2 > 100 and middle_result2 < 1000):
                    total = total + 1



fout.write (str(total) + '\n')
fout.close()